### PR TITLE
'Use left and right keys for page turning' option added

### LINKED ISSUE #12168: Keypad Left/Right navigating chapters, not pages?
* KOReader version: 2024.07
* Device: Kindle 4

#### Issue
I read the release notes and PR comments, and it sounds like the Kindle 4 keypad, which I use all the time while reading one-handed, is being repurposed to navigate chapters rather than pages? Is this true, and can this behaviour be reverted? I can see it's useful, but I navigate pages with left/right. Wouldn't mind remapping up/down to navigate chapters, or holding center button as a shift modifier and pressing left/right to navigate chapters, but I'm holding off on updating KOReader for the moment, as this will break it for me.

#### Steps to reproduce
https://github.com/koreader/koreader/pull/11749#issuecomment-2229527541

### PR DESCRIPTION
**Pull request by Commodore64user**
Rewrite of `registerKeyEvents` in both reader paging and rolling. Now setting `useDPadAsActionKeys() = false` reverts to previous key mapping (c. 2024.04). See #11887, #11900 and https://github.com/koreader/koreader/issues/12168#issuecomment-2231692603

Adds <kbd>spacebar</kbd> as `PgFwd` and removes need to use modifiers for `panning` when in continuous mode. (#11749)

Adds option to turn pages with <kbd>Left</kbd> and <kbd>Right</kbd> keys on devices with `useDPadAsActionKeys() = true`. closes #12168

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12174)

### CODE REVIEW

File: frontend/ui/elements/physical_buttons.lua [hunk 1]
```diff
@@ -49,6 +49,17 @@ if Device:hasDPad() and Device:useDPadAsActionKeys() then
             G_reader_settings:flipNilOrFalse("input_invert_right_page_turn_keys")
             Device:invertButtonsRight()
         end,
+        separator = true,
+    })
+    table.insert(PhysicalButtons.sub_item_table, {
+        text = _("Use left and right keys for page turning"),
+        checked_func = function()
+            return G_reader_settings:isTrue("left_right_turn_pages")
```
Comment (poire-z) [inline frontend/ui/elements/physical_buttons.lua:57]: The other setting names in this file have the word "key" in them. This one should too.

File: frontend/apps/reader/modules/readerpaging.lua [hunk 2]
```diff
@@ -52,99 +52,32 @@ function ReaderPaging:onGesture() end
...
-            { "Right" },
-            event = "GotoNextChapter",
-            args = 1,
-        }
-        self.key_events.GotoPrevChapter = {
-            { "Left" },
-            event = "GotoPrevChapter",
-            args = -1,
-        }
+        if G_reader_settings:isTrue("left_right_turn_pages") then
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerpaging.lua:55]:
```suggestion
        if G_reader_settings:isTrue("left_right_keys_turn_pages") then
```

File: frontend/apps/reader/modules/readerpaging.lua [hunk 3]
```diff
@@ -52,99 +52,32 @@ function ReaderPaging:onGesture() end
...
+            self.key_events.GotoNextPos = { { { "RPgFwd", "LPgFwd", "Right" } }, event = "GotoPosRel", args = 1, }
+            self.key_events.GotoPrevPos = { { { "RPgBack", "LPgBack", "Left" } }, event = "GotoPosRel", args = -1, }
+        elseif G_reader_settings:nilOrFalse("left_right_turn_pages") then
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerpaging.lua:58]:
```suggestion
        elseif G_reader_settings:nilOrFalse("left_right_keys_turn_pages") then
```

File: frontend/apps/reader/modules/readerrolling.lua [hunk 4]
```diff
@@ -116,127 +116,42 @@ end
...
-            { "Right" },
-            event = "GotoNextChapter",
-            args = 1,
-        }
-        self.key_events.GotoPrevChapter = {
-            { "Left" },
-            event = "GotoPrevChapter",
-            args = -1,
-        }
+        if G_reader_settings:isTrue("left_right_turn_pages") then
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerrolling.lua:120]:
```suggestion
        if G_reader_settings:isTrue("left_right_keys_turn_pages") then
```

File: frontend/apps/reader/modules/readerrolling.lua [hunk 5]
```diff
@@ -116,127 +116,42 @@ end
...
+            self.key_events.GotoNextView = { { { "RPgFwd", "LPgFwd", "Right" } }, event = "GotoViewRel", args = 1, }
+            self.key_events.GotoPrevView = { { { "RPgBack", "LPgBack", "Left" } }, event = "GotoViewRel", args = -1, }
+        elseif G_reader_settings:nilOrFalse("left_right_turn_pages") then
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerrolling.lua:123]:
```suggestion
        elseif G_reader_settings:nilOrFalse("left_right_keys_turn_pages") then
```

File: frontend/ui/elements/physical_buttons.lua (hunk 1 above)
Comment (Commodore64user) [inline frontend/ui/elements/physical_buttons.lua:57]:
```suggestion
            return G_reader_settings:isTrue("left_right_keys_turn_pages")
```

File: frontend/ui/elements/physical_buttons.lua [hunk 6]
```diff
@@ -49,6 +49,17 @@ if Device:hasDPad() and Device:useDPadAsActionKeys() then
...
+        end,
+        callback = function()
+            G_reader_settings:flipNilOrFalse("left_right_turn_pages")
```
Comment (Commodore64user) [inline frontend/ui/elements/physical_buttons.lua:60]:
```suggestion
            G_reader_settings:flipNilOrFalse("left_right_keys_turn_pages")
```

**Review by Commodore64user — COMMENTED** [review]
Add word "keys" to `G_settings`

**Comment by Frenzie** [comment]
The rationale behind this PR seems to be mistaken unfortunately? A setting may or may not be desired, but it doesn't seem that all of these big changes are necessary, which is unfortunate since you spent the time. But I'll check later; a bit of a refactor may make things clearer regardless. :-)

**Comment by Commodore64user** [comment]
I would like to make a slight change in behaviour, currently panning (on continuous mode) is under `Shift` + `PgFwd`. I now would like to have just under `PgFwd` and add a similar "space bar as PgFwd" setting:

```lua
if Device:hasSymKey() and G_reader_settings:isTrue("spacebar_as_page_forward") then
    self.key_events.MoveUp = { { { "RPgBack", "Space" } }, event = "Panning", args = {0, -1}, }
    self.key_events.MoveDown = { { "RPgFwd" }, event = "Panning", args = {0,  1}, }
elseif Device:hasScreenKB() or Device:hasSymKey() then
    self.key_events.MoveUp = { { "RPgBack" }, event = "Panning", args = {0, -1}, }
    self.key_events.MoveDown = { { "RPgFwd" }, event = "Panning", args = {0,  1}, }
end 
```

**Comment by Frenzie** [comment]
Sounds sensible enough. Does it need a setting though?

**Comment by Commodore64user** [comment]
> Sounds sensible enough. Does it need a setting though?

I am not against always being active, but perhaps is more responsible to let people decide, not sure. Whatever you think best.

**Comment by Frenzie** [comment]
I'd just add it. It's fairly standard reader behavior. Customizing it can wait until there's demand, or if someone implements a keyboard shortcut manager.

**Comment by Commodore64user** [comment]
Ready to merge.

**Comment by Commodore64user** [comment]
> The rationale behind this PR seems to be mistaken unfortunately? A setting may or may not be desired, but it doesn't seem that all of these big changes are necessary, which is unfortunate since you spent the time. But I'll check later; a bit of a refactor may make things clearer regardless. :-)

I just saw this, not sure what's going on 🙄

File: frontend/ui/elements/physical_buttons.lua [hunk 7]
```diff
@@ -49,6 +49,17 @@ if Device:hasDPad() and Device:useDPadAsActionKeys() then
...
+        separator = true,
+    })
+    table.insert(PhysicalButtons.sub_item_table, {
+        text = _("Use left and right keys for page turning"),
+        checked_func = function()
+            return G_reader_settings:isTrue("left_right_keys_turn_pages")
+        end,
+        callback = function()
+            G_reader_settings:flipNilOrFalse("left_right_keys_turn_pages")
+            UIManager:askForRestart()
```
Comment (Frenzie) [inline frontend/ui/elements/physical_buttons.lua:61]: No restart needed, just something that triggers registerKeyEvents().

That's the `PhysicalKeyboard(Dis)Connected` event, but pinging @NiLuJe in case of any insight/suggestions.

**Comment by Frenzie** [comment]
> I just saw this, not sure what's going on 🙄

Same as in the thread, nothing to worry about

File: frontend/apps/reader/modules/readerrolling.lua [hunk 8]
```diff
@@ -116,127 +116,37 @@ end
...
-            { "ScreenKB", "RPgBack" },
-            event = "Panning",
-            args = {0, -1},
-        }
-        self.key_events.MoveDown = {
-            { "ScreenKB", "RPgFwd" },
-            event = "Panning",
-            args = {0,  1},
-        }
+    if Device:hasKeys() and not (Device:hasScreenKB() or Device:hasSymKey()) then
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerrolling.lua:135]:
```suggestion
    if Device:hasKeys() and not Device:useDPadAsActionKeys() then
```

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerrolling.lua:135]: @Frenzie right?

Comment (Frenzie) [reply frontend/apps/reader/modules/readerrolling.lua:135]: In that case it seems it should at least be something like `if (Device:hasDPad() and not Device:useDPadAsActionKeys()) or (Device:hasKeys() and not Device:useDPadAsActionKeys())` if there isn't an obvious better way to do it. 

Duplicating a few lines is only to be avoided as a rule of thumb btw. Sometimes things get much muddier when trying to avoid repeating things.

Comment (Commodore64user) [inline frontend/apps/reader/modules/readerrolling.lua:135]:
```suggestion
    if (Device:hasDPad() and not Device:useDPadAsActionKeys()) or (Device:hasKeys() and not Device:useDPadAsActionKeys()) then
```

File: frontend/ui/elements/physical_buttons.lua (hunk 7 above)
Comment (Commodore64user) [reply frontend/ui/elements/physical_buttons.lua:61]: okay, only this missing... everything else is ready

**Review by NiLuJe — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12174)* status: all files reviewed, 10 unresolved discussions (waiting on @Commodore64user, @Frenzie, and @poire-z)

___
*[`frontend/ui/elements/physical_buttons.lua` line 61 at r3](https://reviewable.io/reviews/koreader/koreader/12174#gh-1680900263:-O21o_GA8hH0Gx97F9DI:b-7xphmo) ([raw file](https://github.com/koreader/koreader/blob/35624128a9dd1519dd734676c33b3b3d6d2f71cd/frontend/ui/elements/physical_buttons.lua#L61)):*
<details><summary><i>Previously, Commodore64user (David) wrote…</i></summary><blockquote>

okay, only this missing... everything else is ready
</blockquote></details>

We don't really have anything else that wraps `registerKeyEvents` besides those two events, indeed, so, yeah, let's use them with a comment, saves adding another alias just for semantics' sake ;).

(i.e., I want future people working on PhysicalKeyboard events to easily be able to cross-ref that comment via a grep).

**Comment by NiLuJe** [comment]
> let's use them with a comment

You'll probably want to mimic what ExternalKeyboard does, e.g., a broadcast.

**Comment by NiLuJe** [comment]
Wait, ixnay on that.

That will confuse the hell out of FocusManager.

So, no, yeah, either leave it as a restart (which I'm fine with, nobody's going to be swapping this ever 5s), or you'll have to pepper a bunch of event handler aliases with a new name everywhere.

**Review by NiLuJe — APPROVED** [review]
TL;DR: :lgtm:

> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12174#-:-O21pzi9E4rxKAogOH3o:bsr8r8e)* status: all files reviewed, 10 unresolved discussions (waiting on @Commodore64user, @Frenzie, and @poire-z)
